File: src/Azdo.Tui/Components/VotePicker.cs
using Azdo.Tui.Rendering;
using Azdo.Tui.Runtime;
using StyleSet = Azdo.Tui.Styles.Styles;

namespace Azdo.Tui.Components;

/// <summary>Emitted when a vote option is selected (≈ <c>components.VoteSelectedMsg</c>).</summary>
public sealed record VoteSelectedMsg(int Vote) : IMsg;

/// <summary>
/// Azure DevOps PR vote values. Kept local to avoid a dependency on Core's
/// <c>azdevops</c> constants (values must match the API).
/// </summary>
public static class VoteValues
{
    public const int Approve = 10;
    public const int ApproveWithSuggestions = 5;
    public const int NoVote = 0;
    public const int WaitForAuthor = -5;
    public const int Reject = -10;
}

/// <summary>A single vote choice (≈ <c>components.VoteOption</c>).</summary>
public sealed class VoteOption
{
    public string Label { get; init; } = "";
    public string Icon { get; init; } = "";
    public int Vote { get; init; }
}

/// <summary>A modal for selecting a PR vote (≈ <c>components.VotePicker</c>).</summary>
public sealed class VotePicker
{
    private const int MinModalWidth = 80;

    private readonly StyleSet _styles;
    private bool _visible;
    private int _width;
    private int _height;
    private readonly List<VoteOption> _options;
    private int _cursor;

    public VotePicker(StyleSet styles)
    {
        _styles = styles;
        _options = new List<VoteOption>
        {
            new() { Label = "Approve", Icon = "✓", Vote = VoteValues.Approve },
            new() { Label = "Approve with suggestions", Icon = "~", Vote = VoteValues.ApproveWithSuggestions },
            new() { Label = "Wait for author", Icon = "◐", Vote = VoteValues.WaitForAuthor },
            new() { Label = "Reject", Icon = "✗", Vote = VoteValues.Reject },
            new() { Label = "Reset feedback", Icon = "○", Vote = VoteValues.NoVote },
        };
    }

    public void Show() => _visible = true;
    public void Hide() => _visible = false;
    public bool IsVisible => _visible;
    public void SetSize(int width, int height) { _width = width; _height = height; }
    public int GetCursor() => _cursor;

    public Cmd? Update(IMsg msg)
    {
        if (!_visible) return null;
        if (msg is KeyMsg key)
        {
            switch (key.Key)
            {
                case "esc":
                case "q":
                    _visible = false;
                    return null;
                case "up":
                case "k":
                    if (_cursor > 0) _cursor--;
                    return null;
                case "down":
                case "j":
                    if (_cursor < _options.Count - 1) _cursor++;
                    return null;
                case "enter":
                    var selected = _options[_cursor];
                    _visible = false;
                    return Commands.Of(new VoteSelectedMsg(selected.Vote));
            }
        }
        return null;
    }

    public string View()
    {
        if (!_visible) return "";

        var theme = _styles.Theme;
        const string titleText = "Vote on Pull Request";
        const string helpTextStr = "↑/↓: navigate • enter: select • esc/q: cancel";

        int maxWidth = MinModalWidth;
        if (titleText.Length > maxWidth) maxWidth = titleText.Length;
        if (helpTextStr.Length > maxWidth) maxWidth = helpTextStr.Length;

        foreach (var opt in _options)
        {
            int lineLen = $"> {opt.Icon} {opt.Label}".Length;
            if (lineLen > maxWidth) maxWidth = lineLen;
        }

        var optionList = new System.Text.StringBuilder();
        for (int i = 0; i < _options.Count; i++)
        {
            var opt = _options[i];
            string cursor = i == _cursor ? ">" : " ";
            string line = $"{cursor} {opt.Icon} {opt.Label}";

            line = i == _cursor
                ? Style.New().Foreground(theme.SelectForeground).Background(theme.SelectBackground).Width(maxWidth).Render(line)
                : Style.New().Foreground(theme.Foreground).Background(theme.Background).Width(maxWidth).Render(line);

            optionList.Append(line).Append('\n');
        }

        string title = Style.New()
            .Foreground(theme.Primary)
            .Background(theme.Background)
            .Bold()
            .Width(maxWidth)
            .Render(titleText);

        string helpText = Style.New()
            .Foreground(theme.ForegroundMuted)
            .Background(theme.Background)
            .Width(maxWidth)
            .Render(helpTextStr);

        string content = Layout.JoinVertical(HAlign.Left, title, "", optionList.ToString(), helpText);

        var modalStyle = Style.New()
            .WithBorder(Border.Rounded)
            .BorderForeground(theme.Border)
            .Padding(1, 2)
            .Background(theme.Background);

        var modal = modalStyle.Render(content);

        if (_width > 0 && _height > 0)
            modal = Layout.Place(_width, _height, HAlign.Center, VAlign.Center, modal);

        return modal;
    }
}

File: reference/internal/ui/components/votepicker.go
package components

import (
	"fmt"

	"github.com/Elpulgo/azdo/internal/azdevops"
	"github.com/Elpulgo/azdo/internal/ui/styles"
	"github.com/charmbracelet/bubbles/key"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
)

// VoteSelectedMsg is sent when a vote option is selected
type VoteSelectedMsg struct {
	Vote int
}

// VoteOption represents a single vote choice
type VoteOption struct {
	Label string
	Icon  string
	Vote  int
}

// VotePicker is a modal component for selecting a PR vote
type VotePicker struct {
	styles  *styles.Styles
	visible bool
	width   int
	height  int
	options []VoteOption
	cursor  int
}

// NewVotePicker creates a new vote picker
func NewVotePicker(s *styles.Styles) VotePicker {
	return VotePicker{
		styles:  s,
		visible: false,
		options: []VoteOption{
			{Label: "Approve", Icon: "✓", Vote: azdevops.VoteApprove},
			{Label: "Approve with suggestions", Icon: "~", Vote: azdevops.VoteApproveWithSuggestions},
			{Label: "Wait for author", Icon: "◐", Vote: azdevops.VoteWaitForAuthor},
			{Label: "Reject", Icon: "✗", Vote: azdevops.VoteReject},
			{Label: "Reset feedback", Icon: "○", Vote: azdevops.VoteNoVote},
		},
		cursor: 0,
	}
}

// Show makes the vote picker visible
func (v *VotePicker) Show() {
	v.visible = true
}

// Hide makes the vote picker invisible
func (v *VotePicker) Hide() {
	v.visible = false
}

// IsVisible returns whether the vote picker is visible
func (v VotePicker) IsVisible() bool {
	return v.visible
}

// SetSize sets the dimensions for centering
func (v *VotePicker) SetSize(width, height int) {
	v.width = width
	v.height = height
}

// GetCursor returns the current cursor position
func (v VotePicker) GetCursor() int {
	return v.cursor
}

// Update handles messages
func (v VotePicker) Update(msg tea.Msg) (VotePicker, tea.Cmd) {
	if !v.visible {
		return v, nil
	}

	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch {
		case key.Matches(msg, key.NewBinding(key.WithKeys("esc", "q"))):
			v.visible = false
			return v, nil

		case key.Matches(msg, key.NewBinding(key.WithKeys("up", "k"))):
			if v.cursor > 0 {
				v.cursor--
			}
			return v, nil

		case key.Matches(msg, key.NewBinding(key.WithKeys("down", "j"))):
			if v.cursor < len(v.options)-1 {
				v.cursor++
			}
			return v, nil

		case key.Matches(msg, key.NewBinding(key.WithKeys("enter"))):
			selected := v.options[v.cursor]
			v.visible = false
			return v, func() tea.Msg {
				return VoteSelectedMsg{Vote: selected.Vote}
			}
		}
	}

	return v, nil
}

// View renders the vote picker
func (v VotePicker) View() string {
	if !v.visible {
		return ""
	}

	titleText := "Vote on Pull Request"
	helpTextStr := "↑/↓: navigate • enter: select • esc/q: cancel"

	maxWidth := minModalWidth
	if len(titleText) > maxWidth {
		maxWidth = len(titleText)
	}
	if len(helpTextStr) > maxWidth {
		maxWidth = len(helpTextStr)
	}

	for _, opt := range v.options {
		lineLen := len(fmt.Sprintf("> %s %s", opt.Icon, opt.Label))
		if lineLen > maxWidth {
			maxWidth = lineLen
		}
	}

	var optionList string
	for i, opt := range v.options {
		cursor := " "
		if i == v.cursor {
			cursor = ">"
		}

		line := fmt.Sprintf("%s %s %s", cursor, opt.Icon, opt.Label)

		if i == v.cursor {
			line = lipgloss.NewStyle().
				Foreground(v.styles.Theme.GetSelectForeground()).
				Background(v.styles.Theme.GetSelectBackground()).
				Width(maxWidth).
				Render(line)
		} else {
			line = lipgloss.NewStyle().
				Foreground(v.styles.Theme.GetForeground()).
				Background(v.styles.Theme.GetBackground()).
				Width(maxWidth).
				Render(line)
		}

		optionList += line + "\n"
	}

	title := lipgloss.NewStyle().
		Foreground(v.styles.Theme.GetPrimary()).
		Background(v.styles.Theme.GetBackground()).
		Bold(true).
		Width(maxWidth).
		Render(titleText)

	helpText := lipgloss.NewStyle().
		Foreground(v.styles.Theme.GetForegroundMuted()).
		Background(v.styles.Theme.GetBackground()).
		Width(maxWidth).
		Render(helpTextStr)

	content := lipgloss.JoinVertical(
		lipgloss.Left,
		title,
		"",
		optionList,
		helpText,
	)

	modalStyle := lipgloss.NewStyle().
		Border(lipgloss.RoundedBorder()).
		BorderForeground(v.styles.Theme.GetBorder()).
		Padding(1, 2).
		Background(v.styles.Theme.GetBackground())

	modal := modalStyle.Render(content)

	if v.width > 0 && v.height > 0 {
		modal = lipgloss.Place(
			v.width,
			v.height,
			lipgloss.Center,
			lipgloss.Center,
			modal,
		)
	}

	return modal
}
